ARTEMIS-3243 Improve Mirror with dual mirror

clebertsuconic opened the pull request.

clebertsuconic commented:
I am still working on this, there are a few failures.. and I have to do some tweaking when a node is reconnecting... and I will also include duplicate detection as part of this PR... which I should finish by early next week.

clebertsuconic commented:
The idea with this PR is that two nodes could have their mirrors enabled... you just switch the consumers and producers from one node to the next and all the data should be there.

gtully commented on artemis-journal/src/main/java/org/apache/activemq/artemis/core/journal/impl/JournalImpl.java:
```diff
@@ -984,6 +984,7 @@ public void appendDeleteRecord(final long id, final boolean sync, final IOComple
       }
 
 
+
```
I guess you can drop this change, one less file touched.

gemmellr reviewed (commented):
I was starting to have a look at this as you mentioned it on the dev list thread about the release, but I've come to realise at least some of the stuff here is already on main so I'm stopping for now as it seems to be those bits ive commented on so far.

Might be good to update the PR with the actual rebased changes, and move it from being a Draft if you think its ready for review/merge. It also looks like it fails to build in the current PR state.

clebertsuconic commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/ByteUtil.java:
```diff
@@ -205,23 +205,49 @@ public static int bytesToInt(byte[] b) {
-      output[offset + 4] = (byte)(x >> 24);
-      output[offset + 5] = (byte)(x >> 16);
-      output[offset + 6] = (byte)(x >>  8);
+      output[offset] = (byte)(x >>> 56);
+      output[offset + 1] = (byte)(x >>> 48);
+      output[offset + 2] = (byte)(x >>> 40);
+      output[offset + 3] = (byte)(x >>> 32);
+      output[offset + 4] = (byte)(x >>> 24);
+      output[offset + 5] = (byte)(x >>> 16);
+      output[offset + 6] = (byte)(x >>>  8);
```
this was a bug originally... >> in Java is signaled.. if you had a negative value things would be pretty much broken.

gemmellr replied:
Its true >> and >>> can differ, and that can make a difference depending on their usage, but I don't believe it does in this case though.

This repeatedly carves off the low order 8 bits using a cast to byte, after shifting the original bits by between 56 and 0 steps right, to incrementally capture the original 64 bits from high to low. It discards the actual shifted value at each intermediate step, operating on the original long for the next shift done (if any).

That seems to mean it will neither access or store any of the high order bits introduced by the shifts. Those are what can be influenced by use of >> vs >>>, which would mean this does exactly the same as it did to begin with.

gemmellr replied:
(To be clear, >>> usage is fine and if the resulting behaviour hasnt altered then there is obviously no problem from the change being made....I made the comment on the change before I later realised it was already pushed, and was mainly just noting it didnt seem like a necessary change, as smaller changes are nice and easier to review and reason about later)

clebertsuconic replied:
@gemmellr it would been broken without the >>>

>> in Java would add further -1 if the value was negative.. totally different from the intention here.

As the byte was really unsigned, we can't use >> here

As a matter of fact that's how Netty does on their libraries as well.


It wasn't really part of this PR.. it was here as supporting my test. I'm updating the branch now.

gemmellr replied:
I still don't see that the way it is used leaves room for a difference, it never seems to store the value or access the bits that the shifting makes a difference to.

Regardless, if you found a defect, add a test for it when you resolve it, makes it clearer what is being fixed and can save discussions like this.

clebertsuconic commented on artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/tools/PrintData.java:
```diff
@@ -167,6 +164,25 @@ public static void printData(File bindingsDirectory, File messagesDirectory, Fil
 
    }
 
+   public static DescribeJournal printMessages(File messagesDirectory, PrintStream out, boolean safe, boolean printRecords, boolean printSurving, boolean reclaimed) {
```
will send a commit with the fix.

clebertsuconic commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/ByteUtil.java:
```diff
@@ -233,6 +233,21 @@ public static void longToBytes(long x, byte[] output, int offset) {
    }
 
+   /** the byte ID goes to the first byte, everything else stays where they are. */
+   public static long mixByteAndLong(byte id, long longID) {
```
caller's responsibility.

We validate if the ID belong to the server's before we send it.

gemmellr replied:
I'm not talking about if it belongs to a server, but whether this method overwrites any data in the high order bits of the ID when munging the two values together.

clebertsuconic commented on artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPStandardMessage.java:
```diff
@@ -322,17 +323,26 @@ public String toString() {
       ensureScanning();
-      return super.toString();
+      String body = getStringBody();
+      return super.toString() + (body != null ? ",body=[" + body + "]" : "");
```
this is for print-data and debug. It's been quite useful.

clebertsuconic replied:
although.. I will remove it... if we need it in the future I may revisit this.

clebertsuconic commented on artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPStandardMessage.java:
```diff
@@ -322,17 +323,26 @@ public String toString() {
+      if (body instanceof AmqpValue) {
+         AmqpValue sectionBody = (AmqpValue) body;
+         if (sectionBody != null) {
+            if (sectionBody.getValue() instanceof Long) {
+               long id = (long)sectionBody.getValue();
+               return sectionBody.getValue().toString() + ", byte[0]=" + ByteUtil.getFirstByte(id) + ", messageID=" + ByteUtil.removeFirstByte(id);
+            } else {
+               return sectionBody.getValue().toString();
+            }
```
this one was not intentional.. I will remove it.

clebertsuconic commented on artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPBrokerConnection.java:
```diff
@@ -490,12 +495,18 @@ private void connectSender(Queue queue,
             Source source = new Source();
             source.setAddress(queue.getAddress().toString());
             sender.setSource(source);
+            HashMap<Symbol, Object> mapProperties = new HashMap<>(1);
```
@gemmellr I'm only using a single element here. as I read the code, when you call putVal on hashMap, it will resize for 1 element, which is my intention.

gemmellr replied:
Try debugging it. Unless its changed, it will resize itself again immediately during/after insertion due to exceeding the load-factor.

clebertsuconic commented on artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPBrokerConnection.java:
```diff
@@ -490,12 +495,18 @@ private void connectSender(Queue queue,
             source.setAddress(queue.getAddress().toString());
             sender.setSource(source);
+            HashMap<Symbol, Object> mapProperties = new HashMap<>(1);
+            mapProperties.put(AMQPMirrorControllerSource.BROKER_ID, server.getMirrorBrokerId());
```
I'm changing the datatype to byte all the way.

clebertsuconic commented:
@gemmellr I have updated the PR with my suggestions, also I changed the schema to include the target-broker-id and source-broker-id as part of the XML... 

I was stuck with the smoke tests missing acks after failures... and they are now fixed...


I will convert this as a regular PR now.. there are probably some more reviews to be made.. and I will still produce more fixes next week that can come on a different PR / improvements.

clebertsuconic commented:
I will close this PR, do some work and open a new one.

clebertsuconic commented:
there is a new PR for this: #3633
